src/transpose_z_to_y.f90

```diff
@@ -63,94 +61,49 @@ module subroutine transpose_z_to_y_real_long(src, dst, decomp)
-      end if
-#else
-      ! note the src array is suitable to be a send buffer
-      ! so no split operation needed
-
+      nsize = product(decomp%zsz)
 #if defined(_GPU)
-      !$acc host_data use_device(src)
-      istat = cudaMemcpy(wk1, src, s1 * s2 * s3, cudaMemcpyDeviceToDevice)
+      !$acc host_data use_device(src,dst)
+      istat = cudaMemcpy(dst, src, nsize, cudaMemcpyDeviceToDevice)
       !$acc end host_data
-      if (istat /= 0) call decomp_2d_abort(__FILE__, __LINE__, istat, "cudaMemcpy2D")
+      if (istat /= 0) call decomp_2d_abort(__FILE__, __LINE__, istat, "cudaMemcpy")
+#else
+      dst = src
 #endif
 
-#endif
+   end subroutine transpose_z_to_y_real_fast
+   
+   subroutine transpose_z_to_y_real(src, dst, decomp, wk1, wk2)
 
-      ! define receive buffer
-#ifdef EVEN
-      if (decomp%even) then
-         call MPI_ALLTOALL(src, decomp%z2count, real_type, &
-                           wk2, decomp%y2count, real_type, &
-                           DECOMP_2D_COMM_ROW, ierror)
-      else
-         call MPI_ALLTOALL(wk1, decomp%z2count, real_type, &
-                           wk2, decomp%y2count, real_type, &
-                           DECOMP_2D_COMM_ROW, ierror)
-      end if
-      if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_ALLTOALL")
-#else
+      implicit none
 
+      real(mytype), dimension(:, :, :), intent(IN) :: src
+      real(mytype), dimension(:, :, :), intent(OUT) :: dst
+      TYPE(DECOMP_INFO), intent(IN) :: decomp
+      real(mytype), dimension(:), intent(out) :: wk1, wk2
 #if defined(_GPU)
-#if defined(_NCCL)
-      call decomp_2d_nccl_send_recv_row(wk2, &
-                                        wk1, &
-                                        decomp%z2disp, &
-                                        decomp%z2cnts, &
-                                        decomp%y2disp, &
-                                        decomp%y2cnts, &
-                                        dims(2))
-#else
-      call MPI_ALLTOALLV(wk1, decomp%z2cnts, decomp%z2disp, real_type, &
-                         wk2, decomp%y2cnts, decomp%y2disp, real_type, &
-                         DECOMP_2D_COMM_ROW, ierror)
-      if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_ALLTOALLV")
-#endif
-#else
-      associate (wk => wk1)
-      end associate
-      call MPI_ALLTOALLV(src, decomp%z2cnts, decomp%z2disp, real_type, &
-                         wk2, decomp%y2cnts, decomp%y2disp, real_type, &
-                         DECOMP_2D_COMM_ROW, ierror)
-      if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_ALLTOALLV")
-#endif
+      attributes(device) :: wk1, wk2
 #endif
 
-      ! rearrange receive buffer
-      call mem_merge_zy_real(wk2, d1, d2, d3, dst, dims(2), &
-                             decomp%y2dist, decomp)
+      call rearrange_sendbuf_real(src, wk1, decomp)
+      call define_recvbuf_real(src, decomp, wk1, wk2)
```

Comment:
The two other names are good and clear. This one is not so clear in my opinion. What about `perform_transpose_real` ? `sendbuf_to_recvbuf` ?

Reply:
Agreed, I was just going on how the comments described this step

Reply:
Done